src/python/gudhi/representations/archipelago.py

```diff
@@ -0,0 +1,127 @@
+# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+# Author(s):       Martin Royer
+
+import copy
+
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+from gudhi.representations.vector_methods import Atol, TopologicalVector, BettiCurve
+
+
+class Archipelago(BaseEstimator, TransformerMixin):
+    """
+    Transformer that dictionary-wraps persistence diagram vectorizers, i.e. objects from gudhi.representations.vector_methods.
+    One provides persistence diagram vectorizers (by way of either `island` or `island_dict`), and the Archipelago object will
+    |fit on| and |transform = vectorize| lists or series of persistence diagrams.
+    The object is sklearn-API consistent.
+
+    Parameters:
+        island: island for populating archipelago, i.e. object to vectorize the target in each homology
+            dimensions. Must be `copy.deepcopy`-able. *Will be ignored if island_list is given*.
+        island_dict: island dict for populating archipelago, i.e. dictionary of objects to vectorize persistence
+            diagrams according to homology_dimensions.
+
+    Examples
+    >>> pdiagram1 = [(0, (0.0, 2.34)), (0, (0.0, 0.956)), (1, (0.536, 0.856)), (2, (1.202, 1.734))]
+    >>> pdiagram2 = [(0, (0.0, 3.34)), (0, (0.0, 2.956)), (1, (0.536, 1.856)), (2, (1.202, 2.734))]
+    >>> pdiagram3 = [(0, (1.0, 4.34)), (0, (2.0, 3.956)), (1, (1.536, 2.856)), (2, (3.202, 4.734))]
+    >>> list_pdiags = [pdiagram1, pdiagram2, pdiagram3]
+    >>> archipelago = Archipelago(island=Atol())
+    >>> archipelago.fit(X=list_pdiags)
+    >>> archipelago.transform(X=list_pdiags)
+    >>> archipelago = Archipelago(island_dict={2: BettiCurve(resolution=4), 0:Atol()})
+    >>> import pandas as pd
+    >>> series_pdiags = pd.Series(list_pdiags)
+    >>> archipelago.set_output(transform="pandas")
+    >>> archipelago.fit(X=series_pdiags)
+    >>> archipelago.transform(X=series_pdiags)
```

Comment:
Looking at https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html, using the new format of persistence diagrams, and ignoring pandas for now, is this equivalent to the following?
```python
from gudhi.representations import BettiCurve, Atol, Archipelago
import numpy as np
from sklearn.compose import ColumnTransformer
Xpdiagram1 = [np.array([[0.0, 2.34], [0.0, 0.956]]), np.array([[0.536, 0.856]]), np.array([[1.202, 1.734]])]
Xpdiagram2 = [np.array([[0.0, 3.34], [0.0, 2.956]]), np.array([[0.536, 1.856]]), np.array([[1.202, 2.734]])]
Xpdiagram3 = [np.array([[1.0, 3.34], [2.0, 2.956]]), np.array([[1.536, 2.856]]), np.array([[3.202, 4.734]])]
Xlist_pdiags = [Xpdiagram1, Xpdiagram2, Xpdiagram3]
ct=ColumnTransformer([("0",Atol(),0),("1",Atol(),1),("2",Atol(),2)])
print(ct.fit_transform(Xlist_pdiags))
ct=ColumnTransformer([("0",Atol(),0),("2",BettiCurve(resolution=4),2)])
print(ct.fit_transform(Xlist_pdiags))
```
I had to do a couple changes in BettiCurve to let this run (I don't know exactly what `not X` is meant to test for)
```diff
@@ -381,7 +381,7 @@
         if not self.is_fitted():
             raise NotFittedError("Not fitted.")
 
-        if not X:
+        if X is None or len(X) == 0:
             X = [np.zeros((0, 2))]
         
         N = len(X)
@@ -408,7 +408,7 @@
 
         return np.array(bettis, dtype=int)[:, 0:-1]
 
-    def fit_transform(self, X):
+    def fit_transform(self, X, y=None):
         """
         The result is the same as fit(X) followed by transform(X), but potentially faster.
         """
```
and I get similar results, but they differ from what Archipelago returns for dimension 0 (maybe it is just related to random generators).

Reply:
Hmm, they differ because you made a mistake in copying it should be `Xpdiagram3 = [np.array([[1.0, 4.34], [2.0, 3.956]]), np.array([[1.536, 2.856]]), np.array([[3.202, 4.734]])]` and then the results are equal indeed.

Also this seems to be working:
`ct.set_output(transform="pandas")
Xdf_pdiags = pd.DataFrame(Xlist_pdiags)
print(ct.fit_transform(Xdf_pdiags))` so this is possibly the end for this PR, with the new pdiags format.

Reply:
So the one thing this PR does better than your proposition is handle dimensions that fail at `fit`, by not registering the key into the `archipelago_` dict.
It seems desirable for now?
Which makes me think a lot of methods from `vector_methods` are not protected against things like 
``` X is None or len(X) == 0```

Reply:
After further discussion with Marc, I believe the better philosophy for gudhi is what he suggests:
- if one asks for vectorizing of some homology dimensions (say 0, 1 and 2) we provide the structure for the dimensions (for instance with ColumnTransformers from this conversation),
-  and because there can be no homology features at some dimensions, if it turns out there is no data to vectorize, we should print a warning (e.g. "no information in dimension d, problems ahead, consider asking for different homology dimensions") but still let it go to crashing, instead of what I was doing which was removing the problematic dimension.

So we can definitely drop this PR now.

We should (somehow?) salvage the changes to vector_methods: default method for Atol so that it can be initialized with empty arguments, and some tweaks that can be generalized to all vector methods such as implementing `get_feature_names_out` so that we can use the `set_output` feature from sklearn.

Reply:
Yes, we should still include the other changes.

For reference, the discussion was mostly about how to handle empty diagrams in some dimensions. If the user asks for persistence in dimensions 0, 1 and 2, but the diagrams in dimension 2 are all empty (especially during `fit`), we have several options:
1. we can throw an exception (possibly just for the vectorizers that do need fitting like Atol, or Landscape without an explicit `sample_range`)
2. we can (print a warning and) continue, and for this dimension `transform` will either (preferably) return a harmless constant, or return nonsense (including NaN).
3. we can (print a warning and) skip that dimension henceforth, `transform` will act as if the user had only asked for dimensions 0 and 1.

@martinroyer chose the last option in this PR. From a pure ML perspective, that option is convenient. A user can ask the pipeline to aggregate the results in several dimensions and not care about the details, if one dimension is useless and the pipeline can automatically ignore it, that user is happy. Ignoring that dimension also has the advantage over returning a constant that it avoids a useless increase in dimension. It also goes well with the dictionary-like pandas.
From a lower level perspective, the last option can be confusing. The output of transform does not have the expected shape (I can ask for 3 landscapes with 5 points each but get a vector of dimension 10 because one dimension was ignored).
Both options have their advantages. Note that this choice can be handled either inside each vectorizer (say `Atol(skip_on_fit_error=True)` for which `fit([])` defines an empty list of centers, and after that `transform` returns a list of empty arrays, which hopefully disappear when ColumnTransformer tries to append them to the rest), or in a wrapper like ColumnTransformer.
In any case, we should check how we handle (lists of) empty diagrams.

Reply:
Adding to that, I believe there are two "empty diagrams" case to discuss, that of `fit` and that of `transform`.

As for `fit` here is potential design number 4: if we hold the view that once someone asks for vectorizing (say in dimensions 0 and 1) we should provide the structure for doing so, I am inclined to go with "provide the structure at all costs", even without data which means for Atol to always yield centers at the end of `fit` - for instance by random generation. So code-wise we could do something like this:
```
    def fit(self, X, y=None, sample_weight=None):
        if not hasattr(self.quantiser, 'fit'):
            raise TypeError("quantiser %s has no `fit` attribute." % (self.quantiser))

        # In fitting we remove infinite birth/death time points so that every center is finite
        X = [dgm[~np.isinf(dgm).any(axis=1), :] for dgm in X if len(dgm)]

        if len(X) < self.quantiser.n_clusters:
            # If no point to fit, let's arbitrarily put centers in [0, 1)
            X += [np.array([[np.random.random(), np.random.random()]]) for _ in range(self.quantiser.n_clusters - len(X))]
```
This would (?) ensure that quantiser has enough points to fit `n_clusters` regardless of entry `X`.

As for `transform`, one convention could be that empty diagrams are thought of as empty measure therefore vectorized as 0 = no mass. This ensures that no error is thrown which would be pretty desirable at `transform` time? For Atol one way of achieving this is to replace empty diagrams with `np.array([[np.inf, np.inf]])`.

Would be glad to hear your opinions on this!

Reply:
I am globally in favor of option (2), that is, return a constant for all data instance if the corresponding persistence diagrams are all empty. I think it matches better with ML practices, where you usually use homology dimensions without thinking too much beforehand if the persistence diagrams make sense or not. If they do not (and thus constants are returned), pipelines will not crash, but will contain redundant, useless constants, that can be spotted a posteriori with post-analysis of the results, as is sometimes done in other contexts.